libs/lib-client/diagram/src/lib/drag-tiles/update-links.ts

```diff
@@ -0,0 +1,95 @@
+import { DiagramGraph, GraphNodeDictionary, NodeType } from '@flogo-web/lib-client/core';
+
+const taskIdFinder = (nodes: GraphNodeDictionary) => {
+  return (id: string) => nodes && nodes[id] && nodes[id].type === NodeType.Task;
+};
+
+/*
+  Case 1. If root element is moved to different place. It will not have a previous parent and it will have previous child.
+          It will have a new parent and may have new child
+  Case 2. If an element is moved as root. It will have parent, it will have child.
+          It will not have a new parent and will have a child
+  Case 3. If an element is moved to last element of a row. It will have parent, it may have child.
+          It will have a new parent in the branch id and it will not have a new child
+  Case 4. And a normal movement from normal place to another normal place
+ */
+
+export function updateLinks(
+  currentGraph: DiagramGraph,
+  movingItemId: string,
+  newParentId: string
+) {
+  const graph = { ...currentGraph };
+  const findTaskId = taskIdFinder(graph.nodes);
+  const allNodes = graph.nodes;
+  const nodeToMove = allNodes[movingItemId];
+  const prevParentId = nodeToMove.parents.pop();
+  const prevChildId = nodeToMove.children.find(findTaskId);
+  const newParentNode = newParentId && allNodes[newParentId];
+  const newChildId = newParentNode
+    ? newParentNode.children.find(findTaskId)
+    : graph.rootId;
+
+  nodeToMove.children = replaceNodeIds(nodeToMove.children, prevChildId, newChildId);
+  if (newParentId) {
```

Comment:
nodeToMove.parents is mutation here, unless you do what I suggested before and clone nodeToMove